Mathlib/Tactic/MoveAdd.lean

```diff
@@ -0,0 +1,427 @@
+    MetaM (List (Expr × Expr)) := do
+  let sums := ← getOps op (← instantiateMVars tgt)
+  let candidates := ← sums.mapM fun (addends, sum) => do
+    let reord := reorderUsing addends.toList instructions
+    let resummed := sumList1 (prepareOp sum) reord
+    if (resummed != sum) then return some (sum, resummed) else return none
+  return (candidates.toList.reduceOption.toArray.qsort
+    (fun x y : Expr × Expr ↦ (y.1.size  ≤ x.1.size))).toList
+
+/-- `permuteExpr op tgt instructions` takes the same input as `rankSums` and returns the
+expression obtained from `tgt` by replacing all `old_sum`s by the corresponding `new_sum`.
+If there were no required changes, then `permuteExpr` reports this in its second factor. -/
+def permuteExpr (op : Name) (tgt : Expr) (instructions : List (Expr × Bool)) :
+    MetaM Expr := do
+  let permInstructions := ← rankSums op tgt instructions
+  if permInstructions == [] then throwError "The goal is already in the required form"
+  let mut permTgt := tgt
+  for (old, new) in permInstructions do
+    permTgt := permTgt.replace (if · == old then new else none)
+  return permTgt
+
+/--  `pairUp L R` takes to lists `L R : List Expr` as inputs.
+It scans the elements of `L`, looking for a corresponding `DefEq` `Expr`ession in `R`.
+If it finds one such element `d`, then it sets the element `d : R` aside, removing it from `R`, and
+it continues with the matching on the remainder of `L` and on `R.erase d`.
+
+At the end, it returns the sublist of `R` of the elements that were matched to some element of `R`,
+in the order in which they appeared in `L`,
+as well as the sublist of `L` of elements that were not matched, also in the order in which they
+appeared in `L`.
+
+Example:
+```lean
+#eval do
+  let L := [mkNatLit 0, (← mkFreshExprMVar (some (mkConst ``Nat))), mkNatLit 0] -- i.e. [0, _, 0]
+  let R := [mkNatLit 0, mkNatLit 0,                                 mkNatLit 1] -- i.e. [0, 1]
+  dbg_trace f!"{(← pairUp L R)}"
+/-  output:
+`([0, 0], [0])`
+the output LHS list `[0, 0]` consists of the first `0` and the `MVarId`.
+the output RHS list `[0]` corresponds to the last `0` in `L`.
+-/
+```
+-/
+def pairUp : List (Expr × Bool × Syntax) → List Expr →
+    MetaM ((List (Expr × Bool)) × List (Expr × Bool × Syntax))
+  | (m::ms), l => do
+    match ← l.findM? (isDefEq · m.1) with
+      | none => let (found, unfound) := ← pairUp ms l; return (found, m::unfound)
+      | some d => let (found, unfound) := ← pairUp ms (l.erase d)
+                  return ((d, m.2.1)::found, unfound)
+  | _, _ => return ([], [])
+
+open Elab.Tactic
+
+/-- `move_oper_simpCtx` is the `Simp.Context` for the reordering internal to `move_oper`.
+To support a new binary operation, extend the list in this definition, so that it contains
+enough lemmas to allow `simp` to close a generic permutation goal for the new binary operation.
+-/
+def move_oper_simpCtx : MetaM Simp.Context := do
+  -- the `simpOnlyBuiltins` are `eq_self` and `iff_self`
```

Comment:
I'd leave this fact to the documentation of `simpOnlyBuiltins` itself, since it seems likely that this comment goes out of date unnoticed.
```suggestion
```